### ISSUE TITLE
bug: tooltip does not work on mobile devices

### ISSUE THREAD

**Issue by jis3r** [huntabyte/shadcn-svelte#1044]
### Describe the bug

The tooltip-component does not work on mobile devices, as there is no hover-event. Instead, the tooltip-trigger should be clickable, so that mobile users can get the information too. But currently the trigger does not react to click-events on mobile.

### Reproduction

Try to use the tooltip on a mobile device. Fastest way to replicate the issue would be to visit https://www.shadcn-svelte.com/docs/components/tooltip

### System Info

```bash
System:
    OS: macOS 14.4.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 189.98 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.6.1 - /opt/homebrew/bin/node
    npm: 10.5.0 - /opt/homebrew/bin/npm
  Browsers:
    Chrome: 124.0.6367.62
    Safari: 17.4.1
  npmPackages:
    @sveltejs/kit: ^2.0.0 => 2.5.1 
    bits-ui: ^0.21.4 => 0.21.4 
    lucide-svelte: ^0.336.0 => 0.336.0 
    svelte: ^4.2.7 => 4.2.11 
    svelte-sonner: ^0.3.11 => 0.3.18 

(but its also on the shadcn-svelte website)
```

### Severity

annoyance

**Comment by shyakadavis** [comment]
Hello;

Tooltips aren't supposed to work on mobile.

Related:

- https://github.com/huntabyte/bits-ui/issues/314
- https://github.com/huntabyte/shadcn-svelte/issues/671

TL;DR
> If you need a tooltip on mobile, then a Popover may better align with what you're trying to accomplish.

**Comment by jis3r** [comment]
Okay, but I see it like @ivands and think, that there are scenarios where you need the click on mobile and the hover on desktop. And as the popover is just feasible with a workaround, as it does not react on hover, I think it would be nice to have the possibility (at least optional) to trigger a tooltip with a click on mobile. 

And regarding the other issues's discussion about "critical" vs "additonal, not necessary" info: why would anyone want to display this only for desktop users and prevent mobile users from accessing the additional info, even if it is not directly necessarry?

**Comment by shyakadavis** [comment]
> I think it would be nice to have the possibility (at least optional) to trigger a tooltip with a click on mobile.

I think you're literally describing a `Popover`.

Tbh, I believe a conditional check of which device a user is on and swapping between `Tooltip` & `Popover` is really simple. (Am I missing something?)

Plus, as Hunter clearly puts it, on a technical level:

> there is no sense of "hover" on mobile, which is the only time tooltips should be active.

**Comment by jis3r** [comment]
Thanks for the clarification @huntabyte :)

I just worked with the popover and after trying different things, I end up where @isaacharrisholt was and would also love the addition of "hoverability", as mentioned in [issue 395 in Bits UI](https://github.com/huntabyte/bits-ui/issues/395).

Anyways, thanks for the help and keep up the good work!

**Comment by jis3r** [comment]
I found [this neat package](https://github.com/rozek/svelte-device-info) that makes the device-detection relatively easy. 🙌🏼

**Comment by calebvetter** [comment]
Adding my solution here cause this was the first Google hit. I'm using Svelte 5, so this is based on the ["next" version of shadcn-svelte](https://next.shadcn-svelte.com)

This uses the new [MediaQuery](https://svelte.dev/docs/svelte/svelte-reactivity#MediaQuery) reactive class to determine if it's a touch device or not (or more specifically, if it has a hover feature from a pointing device like a mouse).

The general idea is to create the main three components (Root, Trigger, Content) and serve the correct ones based on touch device, matching a few of the stylings in the process, while also passing on all props and classes for full customizability.

**NOTE:** For the Tooltip part to work you _must_ have a `<Tooltip.Provider>` somewhere as the ancestor (see Provider Component [here](https://next.bits-ui.com/docs/components/tooltip)).

```svelte
<!-- /src/lib/components/ui/popover-tooltip/popover-tooltip.svelte -->
<script lang="ts">
	import { MediaQuery } from 'svelte/reactivity';
	import type { PopoverRootProps, TooltipRootProps } from 'bits-ui';
	import { Popover } from '$lib/components/ui/popover';
	import { Tooltip } from '$lib/components/ui/tooltip';

	let { ...restProps }: PopoverRootProps & TooltipRootProps = $props();

	const isTouch = new MediaQuery('any-hover: none', false);
</script>

{#if isTouch.current}
	<Popover {...restProps}>
		{@render restProps.children?.()}
	</Popover>
{:else}
	<Tooltip delayDuration={0} {...restProps}>
		{@render restProps.children?.()}
	</Tooltip>
{/if}
```

```svelte
<!-- /src/lib/components/ui/popover-tooltip/popover-tooltip-trigger.svelte -->
<script lang="ts">
	import { MediaQuery } from 'svelte/reactivity';
	import type { PopoverTriggerProps, TooltipTriggerProps } from 'bits-ui';
	import { PopoverTrigger } from '$lib/components/ui/popover';
	import { TooltipTrigger } from '$lib/components/ui/tooltip';

	let { ...restProps }: PopoverTriggerProps & TooltipTriggerProps = $props();

	const isTouch = new MediaQuery('any-hover: none', false);
</script>

{#if isTouch.current}
	<PopoverTrigger {...restProps}>
		{@render restProps.children?.()}
	</PopoverTrigger>
{:else}
	<TooltipTrigger {...restProps}>
		{@render restProps.children?.()}
	</TooltipTrigger>
{/if}
```

```svelte
<!-- /src/lib/components/ui/popover-tooltip/popover-tooltip-content.svelte -->
<script lang="ts">
	import { MediaQuery } from 'svelte/reactivity';
	import type { ClassArray, ClassDictionary } from 'clsx';
	import { Popover as PopoverPrimitive } from 'bits-ui';
	import { Tooltip as TooltipPrimitive } from 'bits-ui';
	import { cn } from '$lib/utils';
	import * as Popover from '$lib/components/ui/popover';
	import * as Tooltip from '$lib/components/ui/tooltip';

	type Props = {
		class?: string | ClassArray | ClassDictionary | null | undefined;
	};

	let {
		class: className,
		...restProps
	}: Props & PopoverPrimitive.ContentProps & TooltipPrimitive.ContentProps = $props();

	const isTouch = new MediaQuery('any-hover: none', false);
</script>

{#if isTouch.current}
	<Popover.Content class={cn('w-auto px-3 py-1.5 text-sm', className)} side="top" {...restProps}>
		{@render restProps.children?.()}
	</Popover.Content>
{:else}
	<Tooltip.Content class={cn('font-medium', className)} {...restProps}>
		{@render restProps.children?.()}
	</Tooltip.Content>
{/if}
```

```typescript
// /src/lib/components/ui/popover-tooltip/index.ts
import { default as Root } from './popover-tooltip.svelte';
import { default as Content } from './popover-tooltip-content.svelte';
import { default as Trigger } from './popover-tooltip-trigger.svelte';

export {
	Root,
	Content,
	Trigger,
	//
	Root as PopoverTooltip,
	Content as PopoverTooltipContent,
	Trigger as PopoverTooltipTrigger
};
```

Then to use it:

```svelte
<script lang="ts">
	import * as Tooltip from '$lib/components/ui/tooltip';
	import * as PopoverTooltip from '$lib/components/ui/popover-tooltip';
</script>

<Tooltip.Provider>
	<PopoverTooltip.Root>
		<PopoverTooltip.Trigger>
			Try it!
		</PopoverTooltip.Trigger>
		<PopoverTooltip.Content>
			Very nice, you did it!
		</PopoverTooltip.Content>
	</PopoverTooltip.Root>
</Tooltip.Provider>
```
